### DISCUSSION TITLE
[2023-04-11] GUI Team Demo Day

### DISCUSSION THREAD

**Discussion by wdanilo** [enso-org/enso#6238] [Weekly Demo Day]
Let me start! :)

Unfortunately, I spend a lot of time (3+ days straight) on tracking down an animation bug. It turned out, that when we were creating animations, they were fired 1 frame too late and the bug was hidden in a 2-years old code :( In the end, animations were sometimes jumping if connected with other animations. Now, everything works smoothly and I have very smooth vector editor implementation. It lacks some features still, like adding elements, but it is easy to be extended now and the biggest problems are solved. I'm sad that these problems occurred though :(

https://user-images.githubusercontent.com/1623053/231166699-21f74806-2495-463a-935a-9fb2758752c3.mov

**Comment by Procrat** [comment]
Building on work from the weeks before, I added a node button to enable/disable the output context for that particular node.

I also added a temporary shortcut (cmd-shift-c) to switch the execution environment so we can properly test it.

The "panic" you can see in the video is because the expression it generates isn't recognised by the engine yet.

https://user-images.githubusercontent.com/607786/231182937-738f4f9d-1654-48c5-ad23-91e5f47c25d7.mp4

**Comment by MichaelMauderer** [comment]
I was working on extending the functionality of the execution dropdown menu. The new functionality allows switching the execution environment by keypress and includes the IDE controller logic for sending the information about the selected execution environment to the language server.

**Comment by JaroslavTulach** [comment]
Engine performance related:
- Finished batch insert suggestions. #6189
- Initialization phase (_from first message to visualization delivered_) is now down to **5s**. When we started it was **11s**.
- Investigated issue with `Math.sqrt` slow performance.
- Performance issue with `Any.==` slowness caused by `Panic.catch` fixed. #6184 
-  We finally understand where the problem of `if` versus `case` is. #5709 
- Ability to compare branches with `bench_download` tool. #6198 
- Graphs of performance [available to everyone](https://enso-org.github.io/engine-benchmark-results/)

**Comment by vitvakatu** [comment]
Last week I finished the controllers part of the context switch expressions (https://github.com/enso-org/enso/issues/5927) presented in the last demo. 
For the rest of the week, I was implementing support for the read-only mode in the IDE. The current implementation explicitly checks if the read-only mode is enabled and forbids certain actions on the controllers. It prevents accidental communication with the engine when some GUI bug allows edits in read-only mode. Right now, it leads to various interesting side-effects (like an attempt to edit the node breaks connections between nodes). However, this would not be visible to the user in the future, as forbidden actions would be caught earlier on the View side.

In the video, you can see that attempts to edit the code or project name result in errors in the console. We also do not trigger the evaluation of the code when moving nodes.

https://user-images.githubusercontent.com/6566674/231190389-4eaccd69-2517-4d10-8555-7177b333d8cc.mp4

**Comment by Frizi** [comment]
I was and still am working on widgets on nodes. I have added the display of connected ports, and propagating their on-screen positions to edges. It is not yet possible to edit the connections with drag and drop, but I'm not far from implementing that.

![image](https://user-images.githubusercontent.com/919491/231191323-f6a04f38-3362-46c7-a58b-76afa3dde2f9.png)

**Comment by kazcw** [comment]
Last week:
- #6030: Added GPU debugging feature. Tested GPU-accelerated remote access (**Parsec** works well).
- #6197: Fixed bug. Added unit tests.
- #6231: Started replacing old shaders with universal Rectangle.

Next week, continuing #6231: Add shadow support to Rectangle, and replace more shapes with it.

**Comment by mwu-tow** [comment]
Last week summary:
* Landed the changes adding support for file associations on Windows and Linux.
* Started work on URL deep link support on Windows. Had a design discussions with Wojciech, Sylwia and Paweł. We agreed that we want to support having multiple instances of IDE running side-by-side. This is all ongoing work:
* * I have support for defining URL protocol handler on Windows.
* * I implemented code that recognizes that IDE instance was spawned as a default URL handle.
* * I implemented provisional code that uses Electron's application lock to route the URL data to the correct instance of the IDE.
* * A number of other fixes allowed me to authenticate with GitHub on Windows.
